BaragonAgentService/src/main/java/com/hubspot/baragon/agent/handlebars/PreferSameRackWeightingHelper.java

```diff
@@ -51,4 +50,87 @@ public CharSequence preferSameRackWeighting(Collection<UpstreamInfo> upstreams,
+      if (upstreamInfo.getRackId().isPresent()) {
+        allRacks.add(upstreamInfo.getRackId().get());
+      }
+    }
+    return allRacks;
+  }
+
+  /**
+   *
+   * @param allRacks
+   * @return the capacity that each upstream should handle
+   */
+  private double calculateCapacity(Multiset<String> allRacks) {
+    /* capacity is the proportion of unique upstreams to all the upstreams - this gives the amount of load that each upstreams should carry*/
+    return allRacks.elementSet().size() / (double) allRacks.size();
+  }
+
+  /**
+   *
+   * @param allRacks
+   * @return the total pending load that have to be distributed to other upstreams
+   */
+  private double getTotalPendingLoad(Multiset<String> allRacks) {
+    final double capacity = calculateCapacity(allRacks);
+    double pendingLoad = 0;
+    for (String rack: allRacks) {
+      final double myLoad = 1.0 / allRacks.count(rack);
+      double myPendingLoad = capacity - myLoad;
+      if (myPendingLoad > 0) {
+        pendingLoad += myPendingLoad;
+      }
+    }
+    return pendingLoad;
+  }
+
+  /**
+   *
+   * @param upstreams
+   * @param currentUpstream
+   * @param options
+   * @return the weight of the current upstream relative to the current rack such that each upstream carries an equal load.
+   *
+   * Addressing github issue: https://github.com/HubSpot/Baragon/pull/270
+   * Calculating weights for services such that, even if AZ distribution is uneven among upstreams, they still get an even distribution of traffic
+   */
+  public CharSequence preferSameRackWeightingBalanced(Collection<UpstreamInfo> upstreams, UpstreamInfo currentUpstream, Options options) {
+    if (agentMetadata.getEc2().getAvailabilityZone().isPresent() && currentUpstream.getRackId().isPresent()) {
+
+      final String currentRack = agentMetadata.getEc2().getAvailabilityZone().get();
+      final String testingRack = currentUpstream.getRackId().get();
+      final Multiset<String> allRacks = generateAllRacks(upstreams);
+
+      final double capacity = calculateCapacity(allRacks);
+      final double load = 1.0 / allRacks.count(currentRack);
+
+      if (currentRack.equals(testingRack)) {
+        if (load < capacity) { return ""; }
+        return String.format(configuration.getWeightingFormat(), (int) Math.ceil(capacity * allRacks.size()));
+      }
+
+      final double pendingLoadInCurrentRack = load - capacity;
+      if (pendingLoadInCurrentRack <= 0) { return configuration.getZeroWeightString(); }
+
+      final double extraCapacityInTestingRack = capacity - (1.0 / (allRacks.count(testingRack)));
+      if (extraCapacityInTestingRack <= 0) { return configuration.getZeroWeightString(); }
+
+      final double totalPendingLoad = getTotalPendingLoad(allRacks);
+      final double pendingLoadFromCurrentRack = (extraCapacityInTestingRack / totalPendingLoad) * pendingLoadInCurrentRack;
+      final int weight = (int) Math.ceil(pendingLoadFromCurrentRack);
```

Comment:
are there any cases where this rounding will lead to an incorrect answer (i.e. where we'd want a least common denominator type of transformation so that we're always working with ints)? Should we maybe set up a test case with a larger count of upstreams to make sure it comes out as expected? (as opposed to our very simple 1 AZ with too many case)

Reply:
"In Java, double values are IEEE floating point numbers. Unless they are a power of 2 (or sums of powers of 2, e.g. 1/8 + 1/4 = 3/8), they cannot be represented exactly, even if they have high precision. Some floating point operations will compound the round-off error present in these floating point numbers. " 

To resolve this, I've changed the doubles into BigDecimal so to get the precision needed.